provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+'use strict';
+
+var Promise = require('promise');
+var lodash = require('lodash');
+var util = require('util');
+var assert = require('assert');
+var debug = require('debug')('aws-provisioner:aws-manager');
+
+
+/**
+ * AWS EC2 state at a specific moment in time
+ */
+function AwsManager(ec2, keyPrefix, pubKey) {
+  assert(ec2);
+  assert(keyPrefix);
+  assert(pubKey);
+  this.ec2 = ec2;
+  this.keyPrefix = keyPrefix;
+  this.pubKey = pubKey;
+  this.__apiState = {};
+  this.__internalState = {};
+  this.__knownKeyPairs = [];
+}
+
+module.exports = AwsManager;
+
+
+/**
+ * Update the state from the AWS API and return a promise
+ * with no resolution value when completed.
+ */
+AwsManager.prototype.update = function() {
+  var that = this;
+
+  var p = Promise.all([
+    that.ec2.describeInstances({
+      Filters: [{
+        Name: 'key-name',
+        Values: [that.keyPrefix + '*']
+      },{
+        Name: 'instance-state-name',
+        Values: ['running', 'pending']
+      }
+    ]}),
+    that.ec2.describeSpotInstanceRequests({
+      Filters: [{
+        Name: 'launch.key-name',
+        Values: [that.keyPrefix + '*']
+      }, {
+        Name: 'state',
+        Values: ['open']
+      }]
+    }),
+  ]);
+
+  p = p.then(function(res) {
+    that.__apiState = that._classify(res[0], res[1]);
+  });
+
+  p = p.then(function() {
+    // We want to make sure that our internal state is always up to date when
+    // we fetch the updated state
+    that.reconcileInternalState();
+  });
+
+  return p;
+};
+
+
+/**
+ * Classify the state received from AWS into something in the shape:
+ * {
+ *   region: {
+ *     workerTypeName: {
+ *       running: [<Instance>],
+ *       pending: [<Instance>],
+ *       spotReq: [<SpotRequest>],
+ *     }
+ *   }
+ * }
+ * The Instance and SpotRequest objects are those returned by AWS.
+ * We flatten the Reservations because we don't really care about that
+ * feature right now.
+ */
+AwsManager.prototype._classify = function(instanceState, spotReqs) {
+  var that = this;
+  var state = {};
+
+  that.ec2.regions.forEach(function(region) {
+    var rState = state[region] = {};
+
+    function x(type) {
+      if (!rState[type]) {
+        rState[type] = {
+          running: [],
+          pending: [],
+          spotReq: [],
+        };
+      }
+    }
+
+    instanceState[region].Reservations.forEach(function(reservation) {
+      reservation.Instances.forEach(function(instance) {
+        var workerType = instance.KeyName.substr(that.keyPrefix.length);
+        x(workerType);
+        rState[workerType][instance.State.Name].push(instance); 
+      });
+    });
+
+    spotReqs[region].SpotInstanceRequests.forEach(function(request) {
+      var workerType = request.LaunchSpecification.KeyName.substr(that.keyPrefix.length);
+    }); 
+
+  });
+
+  return state;
+};
+
+
+/**
+ * Get Api State Only
```

Comment:
Like this:
```javascript
/**
 * Classify the state received from AWS into something in the shape:
 * {
 *   region: {
 *     workerTypeName: {
 *       running: [<Instance>],
 *       pending: [<Instance>],
 *       spotReq: [<SpotRequest>],
 *     }
 *   }
 * }
 * The Instance and SpotRequest objects are those returned by AWS.
 * We flatten the Reservations because we don't really care about that
 * feature right now.
 */
AwsManager.prototype._classify = function(instanceState, spotReqs) {
```